File: python/fastscapelib/tests/test_flow_snapshot.py
import numpy as np
import numpy.testing as npt
import pytest

from fastscapelib.flow import (
    FlowDirection,
    FlowGraph,
    FlowOperator,
    FlowSnapshot,
    MSTSinkResolver,
    PFloodSinkResolver,
    SingleFlowRouter,
)
from fastscapelib.grid import NodeStatus, RasterGrid


@pytest.fixture
def grid() -> RasterGrid:
    return RasterGrid([10, 20], [1.0, 2.0], NodeStatus.FIXED_VALUE)


def test_flow_snapshot_class_attr() -> None:
    assert FlowSnapshot.graph_updated is False
    assert FlowSnapshot.elevation_updated is False
    assert FlowSnapshot.in_flowdir == FlowDirection.UNDEFINED
    assert FlowSnapshot.out_flowdir == FlowDirection.UNDEFINED


def test_flow_snapshot_constructor() -> None:
    snapshot = FlowSnapshot("test")
    assert isinstance(snapshot, FlowOperator)
    assert snapshot.name == "flow_snapshot"
    assert snapshot.snapshot_name == "test"
    assert snapshot.save_graph is True
    assert snapshot.save_elevation is False

    assert repr(snapshot) == "FlowSnapshot 'test' (graph=True, elevation=False)"


def test_flow_snapshot_error(grid) -> None:
    with pytest.raises(ValueError, match="no flow routing operator defined before"):
        FlowGraph(grid, [FlowSnapshot("test"), SingleFlowRouter()])


def test_flow_snapshot_items(grid) -> None:
    graph = FlowGraph(
        grid,
        [
            PFloodSinkResolver(),
            FlowSnapshot("a", save_graph=False, save_elevation=True),
            SingleFlowRouter(),
            FlowSnapshot("b", save_graph=True, save_elevation=False),
        ],
    )

    assert graph.graph_snapshot_keys == ["b"]
    assert graph.elevation_snapshot_keys == ["a"]

    assert isinstance(graph.graph_snapshot("b"), FlowGraph)
    assert isinstance(graph.elevation_snapshot("a"), np.ndarray)


def test_snapshot_graph(grid) -> None:
    graph = FlowGraph(
        grid,
        [SingleFlowRouter(), FlowSnapshot("a"), MSTSinkResolver(), FlowSnapshot("b")],
    )

    elevation = np.random.uniform(size=grid.shape)
    graph.update_routes(elevation)
    snapshot_a = graph.graph_snapshot("a")
    snapshot_b = graph.graph_snapshot("b")

    attrs = [
        "receivers_count",
        "receivers",
        "receivers_distance",
        "receivers_weight",
        "donors_count",
        "dfs_indices",
    ]

    for name in attrs:
        arr = getattr(graph.impl(), name)
        snapshot_a_arr = getattr(snapshot_a.impl(), name)
        snapshot_b_arr = getattr(snapshot_b.impl(), name)

        assert arr is not snapshot_a_arr

        npt.assert_equal(arr, snapshot_b_arr)
        assert arr is not snapshot_b_arr

    npt.assert_equal(snapshot_b.impl().donors[:, 0], graph.impl().donors[:, 0])
    assert not np.all(snapshot_a.impl().receivers == snapshot_b.impl().receivers)

    # high-level interface test
    actual = graph.accumulate(1.0)
    expected = snapshot_b.accumulate(1.0)
    npt.assert_allclose(actual, expected)

    # test snapshot graphs are read-only
    with pytest.raises(RuntimeError, match=".*(read-only)"):
        snapshot_a.update_routes(elevation)


def test_snapshot_graph_impl(grid) -> None:
    # Ensure that this will not crash Python (segfault). The correct pybind11
    # return value policy for accessing the implementation from a snapshot graph
    # is "reference_internal" (we need the keep_alive call policy to prevent the
    # parent Python graph / snapshot / impl wrappers getting garbage collected
    # when accessing the final graph implementation data members).
    graph = FlowGraph(
        grid,
        [SingleFlowRouter(), FlowSnapshot("a"), MSTSinkResolver(), FlowSnapshot("b")],
    )

    elevation = np.random.uniform(size=grid.shape)
    graph.update_routes(elevation)

    graph.graph_snapshot("a").impl().receivers
    s = graph.graph_snapshot("a")
    s.impl().receivers
    simpl = s.impl()
    simpl.receivers


def test_snapshot_elevation(grid) -> None:
    graph = FlowGraph(
        grid,
        [
            FlowSnapshot("a", save_graph=False, save_elevation=True),
            PFloodSinkResolver(),
            FlowSnapshot("b", save_graph=False, save_elevation=True),
            SingleFlowRouter(),
        ],
    )

    elevation = np.random.uniform(size=grid.shape)
    filled_elevation = graph.update_routes(elevation)
    snapshot_a = graph.elevation_snapshot("a")
    snapshot_b = graph.elevation_snapshot("b")

    npt.assert_equal(elevation, snapshot_a)
    assert elevation is not snapshot_a
    npt.assert_equal(filled_elevation, snapshot_b)
    assert filled_elevation is not snapshot_b

File: test/test_flow_snapshot.cpp
#include "xtensor/xtensor.hpp"
#include "xtensor/xstrided_view.hpp"

#include "gtest/gtest.h"

#include "fastscapelib/flow/flow_graph.hpp"
#include "fastscapelib/flow/flow_router.hpp"
#include "fastscapelib/flow/flow_snapshot.hpp"
#include "fastscapelib/grid/raster_grid.hpp"
#include "fastscapelib/flow/sink_resolver.hpp"


namespace fs = fastscapelib;

namespace fastscapelib
{
    namespace testing
    {

        class flow_snapshot : public ::testing::Test
        {
        protected:
            using flow_graph_type = fs::flow_graph<fs::raster_grid>;
            using grid_type = fs::raster_grid;
            using size_type = typename grid_type::size_type;

            // bottom border base-level
            fs::node_status fixed = fs::node_status::fixed_value;
            fs::node_status core = fs::node_status::core;
            fs::raster_boundary_status bottom_base_level{ { core, core, core, fixed } };

            grid_type grid = grid_type({ 4, 4 }, { 1.0, 1.0 }, bottom_base_level);

            // for the tests here we don't really care about the actual elevation values
            xt::xtensor<double, 2> elevation{ { 0.6, 0.6, 0.6, 0.6 },
                                              { 0.4, 0.4, 0.4, 0.4 },
                                              { 0.2, 0.2, 0.2, 0.2 },
                                              { 0.1, 0.0, 0.1, 0.1 } };
        };

        TEST_F(flow_snapshot, ctor)
        {
            auto snapshot = fs::flow_snapshot("test");
            EXPECT_EQ(snapshot.name(), "flow_snapshot");
            EXPECT_EQ(snapshot.snapshot_name(), "test");
            EXPECT_EQ(snapshot.save_graph(), true);
            EXPECT_EQ(snapshot.save_elevation(), false);
        }

        TEST_F(flow_snapshot, error)
        {
            // no flow routing operator defined before snapshot
            EXPECT_THROW(
                flow_graph_type(grid, { fs::flow_snapshot("a"), fs::single_flow_router() }),
                std::invalid_argument);
        }

        TEST_F(flow_snapshot, items)
        {
            auto graph = flow_graph_type(grid,
                                         {
                                             fs::pflood_sink_resolver(),
                                             fs::flow_snapshot("a", false, true),
                                             fs::single_flow_router(),
                                             fs::flow_snapshot("b", true, false),
                                         });

            ASSERT_EQ(graph.graph_snapshot_keys(), std::vector<std::string>{ "b" });
            ASSERT_EQ(graph.elevation_snapshot_keys(), std::vector<std::string>{ "a" });

            ASSERT_EQ(graph.graph_snapshot("b").size(), grid.size());
            ASSERT_EQ(graph.elevation_snapshot("a").size(), grid.size());
        }

        TEST_F(flow_snapshot, graph)
        {
            auto graph = flow_graph_type(grid,
                                         {
                                             fs::single_flow_router(),
                                             fs::flow_snapshot("a"),
                                             fs::mst_sink_resolver(),
                                             fs::flow_snapshot("b"),
                                         });

            graph.update_routes(elevation);
            auto& snapshot_a = graph.graph_snapshot("a");
            auto& snapshot_b = graph.graph_snapshot("b");

            // no operator in snapshot graphs
            ASSERT_EQ(snapshot_a.operators().size(), 0);
            ASSERT_EQ(snapshot_b.operators().size(), 0);

            ASSERT_NE(&graph, &snapshot_a);
            ASSERT_NE(&graph, &snapshot_b);

            // high-level interface test
            auto actual = graph.accumulate(1.0);
            auto expected = snapshot_b.accumulate(1.0);
            EXPECT_EQ(actual, expected);

            // test snapshot graphs are read-only
            ASSERT_THROW(snapshot_a.update_routes(elevation), std::runtime_error);
            ASSERT_THROW(snapshot_a.set_base_levels(std::vector<size_type>({ 0 })),
                         std::runtime_error);
            ASSERT_THROW(snapshot_a.set_mask(xt::zeros<bool>(grid.shape())), std::runtime_error);
        }

        TEST_F(flow_snapshot, elevation)
        {
            auto graph = flow_graph_type(grid,
                                         {
                                             fs::flow_snapshot("a", false, true),
                                             fs::pflood_sink_resolver(),
                                             fs::flow_snapshot("b", false, true),
                                             fs::single_flow_router(),
                                         });

            const auto& filled_elevation = graph.update_routes(elevation);
            const auto& snapshot_a = graph.elevation_snapshot("a");
            const auto& snapshot_b = graph.elevation_snapshot("b");

            EXPECT_EQ(elevation, snapshot_a);
            EXPECT_EQ(filled_elevation, snapshot_b);
        }
    }

}
